src/workflow/unlock.c

```diff
@@ -118,37 +73,129 @@ keystore_error_t workflow_unlock_and_handle_error(const char* password)
     return unlock_result;
 }
 
-bool workflow_unlock(void)
+static void _unlock_cleanup_password(unlock_data_t* data)
+{
+    if (data->password) {
+        util_zero(data->password, sizeof(data->password));
+        free(data->password);
+        data->password = NULL;
+    }
+}
+
+/**
+ * Callback invoked when the password entry completes.
+ */
+static void _unlock_password_completed(const char* password, void* param)
+{
+    unlock_data_t* data = (unlock_data_t*)param;
+    _unlock_cleanup_password(data);
+    data->password = util_strdup(password);
+    data->state = UNLOCK_STATUS_PW_AVAILABLE;
+}
+
+static void _unlock_init(workflow_t* self)
 {
+    unlock_data_t* data = (unlock_data_t*)self->data;
     if (!memory_is_initialized()) {
-        return false;
+        data->state = UNLOCK_STATUS_FINISHED;
+        data->result = false;
+        return;
     }
     if (!keystore_is_locked()) {
-        return true;
+        data->state = UNLOCK_STATUS_FINISHED;
+        data->result = true;
+        return;
     }
 
+    /*
+     * FUTURE: maybe remove?
+     * It might be a "security" feature/hack, but it's not
```

Comment:
it's not a security feature, it's a "failsafe": 

- some screen is launched and on the stack (e.g. the random number call)
- app quits
- app restarts
- app sends unlock command
- old screen is still there, op_unlock pushes on top of it, and after it's popped, the old screen reappears

pop_all() is/was here because a new app/wallet will call op_unlock as one of the first things.

Not sure if we can get rid of it yet.

Reply:
That can't happen with the current code, can it?
* Blocking workflow (e.g. random number) starts
* App quits
* App restarts
* App sends unlock commands -> USB stack is blocked, command is dropped

Or with the async workflows:
* Async workflow (e.g. random number) starts, locks the USB stack
* App quits
* App restarts
* App sends unlock commands -> USB stack is blocked, command is rejected

And after implementing timers:
* Async workflow (e.g. random number) starts, locks the USB stack
* App quits
* App restarts
* App sends unlock commands -> USB stack is blocked, command is rejected
* Original workflow times out after a bit -> USB stack unlocked
* App retries the unlock command -> Unlock workflow starts.

Reply:
> pop_all() is/was here because a new app/wallet will call op_unlock as one of the first things.
> 
> Not sure if we can get rid of it yet.

It would be great if we never have to do `pop_all()`.